File: kakao/20년 인턴/경주로 건설.py
dx=[0,1,0,-1];dy=[1,0,-1,0]
def solution(board):
    n=len(board);cost=[[0]*n for _ in '1'*n];q=[(0,0,0,1)];q.append((0,0,0,0))
    while q:
        y,x,c,d=q.pop(0)
        for i in range(4):
            ny=y+dy[i];nx=x+dx[i]
            if 0<=ny<n and 0<=nx<n and board[ny][nx]==0 and abs(d-i)!=2:
                nc=c+(100 if d==i else 600)
                if cost[ny][nx]==0 or cost[ny][nx]>=nc:cost[ny][nx]=nc;q.append((ny,nx,nc,i))
    return cost[n-1][n-1]

File: kakao/20년 인턴/경주로 건설.cpp
#include <bits/stdc++.h>

using namespace std;

int cost[25][25][2];
int moveArr[4][2] = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

struct node
{
    int y, x, dir;
};

bool operator>(node a, node b)
{
    return cost[a.y][a.x][a.dir] > cost[b.y][b.x][b.dir];
}

int solution(vector<vector<int>> board) {
    int answer = 0;
    int size = board.size();
    priority_queue < node, vector<node>, greater<node> > pq;

    for (int i = 0; i < 25; i++)
    {
        for (int j = 0; j < 25; j++)
        {
            cost[i][j][0] = 1<<29;
            cost[i][j][1] = 1<<29;
        }
    }

    if (board[0][1] == 0)
    {
        cost[0][1][1] = 100;
        pq.push({ 0, 1, 1 });
    }
    if (board[1][0] == 0)
    {
        cost[1][0][0] = 100;
        pq.push({ 1, 0, 0 });
    }

    while (!pq.empty())
    {
        node curr = pq.top();
        pq.pop();

        for (int i = 0; i < 4; i++)
        {
            int ny = curr.y + moveArr[i][0];
            int nx = curr.x + moveArr[i][1];
            int ndir = i / 2;
            if (ny < 0 || ny >= size || nx < 0 || nx >= size) continue;
            if (board[ny][nx] == 1) continue;
            int newCost = cost[curr.y][curr.x][curr.dir] + ((curr.dir == ndir) ? (100) : (600));
            if (cost[ny][nx][ndir] <= newCost) continue;
            cost[ny][nx][ndir] = newCost;
            pq.push({ny, nx, ndir});
        }
    }

    return min(cost[size - 1][size - 1][0], cost[size - 1][size - 1][1]);
}